src/Plutip/Types.purs

```diff
@@ -197,27 +231,116 @@ instance DecodeAeson StopClusterResponse where
+  keyWallets _ wallet = [ wallet ]
 
 instance
-  UtxoDistribution restSpec restWallets =>
-  UtxoDistribution (InitialUTxO /\ restSpec) (KeyWallet /\ restWallets) where
-  encodeDistribution (amounts /\ rest) =
-    encodeDistribution amounts <> encodeDistribution rest
-  decodeWallets = Array.uncons >>> case _ of
-    Nothing -> Nothing
-    Just { head: PrivateKeyResponse key, tail } ->
-      Tuple (privateKeysToKeyWallet (PrivatePaymentKey key) Nothing) <$>
-        decodeWallets tail
+  ( UtxoDistribution headSpec headWallets
+  , UtxoDistribution restSpec restWallets
+  ) =>
+  UtxoDistribution (headSpec /\ restSpec) (headWallets /\ restWallets) where
+  encodeDistribution (distr /\ rest) =
+    encodeDistribution distr <> encodeDistribution rest
+  decodeWallets (distr /\ rest) =
+    Array.uncons >=>
+      \{ head, tail } -> do
+        wallet <- decodeWallets distr [ head ]
+        Tuple wallet <$> decodeWallets rest tail
+  keyWallets _ (headWallets /\ restWallets) =
+    (keyWallets (Proxy :: Proxy headSpec) headWallets)
+      <> (keyWallets (Proxy :: Proxy restSpec) restWallets)
+
+type WalletInfo =
+  { utxos :: Utxo
+  , payPkh :: PaymentPubKeyHash
+  , stakePkh :: StakePubKeyHash
+  , wallet :: KeyWallet
+  }
+
+-- | For each wallet which includes a stake key, transfer the value of
+-- | the utxos at its enterprise address to its base address.
+transferFundsFromEnterpriseToBase
+  :: forall (r :: Row Type)
+   . PrivatePaymentKey
+  -> Array KeyWallet
+  -> Contract r Unit
+transferFundsFromEnterpriseToBase ourKey wallets = do
+  -- Get all utxos and key hashes at all wallets containing a stake key
+  walletsInfo <- foldM addStakeKeyWalletInfo mempty wallets
+  unless (null walletsInfo) do
+    let ourWallet = privateKeysToKeyWallet ourKey Nothing
+    ourAddr <- liftedM "Could not get our address"
+      $ withKeyWallet ourWallet getWalletAddress
+    ourUtxos <- liftedM "Could not find our utxos" $ utxosAt ourAddr
+    ourPkh <- withKeyWallet ourWallet $ liftedM "Could not get our payment pkh"
+      ownPaymentPubKeyHash
+    let
+      lookups = Lookups.unspentOutputs (unwrap ourUtxos)
+        <> foldMap (_.utxos >>> Lookups.unspentOutputs) walletsInfo
+
+      constraints :: Constraints.TxConstraints Unit Unit
+      constraints = Constraints.mustBeSignedBy ourPkh
+        <> foldMap constraintsForWallet walletsInfo
+    unbalancedTx <- liftedE $ Lookups.mkUnbalancedTx lookups constraints
+    signedTx <- liftedE $ withKeyWallet ourWallet $
+      balanceAndSignTxE unbalancedTx
+    signedTx' <- foldM
+      ( \tx { wallet } -> liftedM "Could not sign" $ withKeyWallet wallet $
+          signTransaction tx
+      )
+      (unwrap signedTx)
+      walletsInfo
+    txHash <- submit (wrap signedTx')
+    awaitTxConfirmed txHash
+  where
+  constraintsForWallet :: WalletInfo -> Constraints.TxConstraints Unit Unit
+  constraintsForWallet { utxos, payPkh, stakePkh } =
+    -- It's necessary to include `mustBeSignedBy`, we get a
```

Comment:
No, this is probably a bug on our side. `mustBeSignedBy` should not be required in any case. Please create a new issue:

- Include a permanent link to this line
- add a link to the issue in the comments here. Use `TODO` keyword